cggmp21/src/signing.rs

```diff
@@ -1160,68 +1273,68 @@ where
     tracer.msgs_received();
 
     tracer.stage("Validate psi_prime_prime");
-    let mut faulty_parties = vec![];
-    for ((j, msg_id, msg_j), (_, ciphertext_id, ciphertext_j)) in
-        round3_msgs.iter_indexed().zip(ciphertexts.iter_indexed())
-    {
-        let R_j = &R[usize::from(j)];
-        let enc_j = fast_paillier::EncryptionKey::from_n(R_j.N.clone());
-
-        let data = pi_log::Data {
-            key0: &enc_j,
-            c: &ciphertext_j.K,
-            x: &msg_j.Delta,
-            b: &Gamma,
-        };
-
-        if pi_log::non_interactive::verify::<E, D>(
-            &unambiguous::ProofLog {
-                sid,
-                prover: j,
-                prime_prime: true,
-            },
-            &R_i.into(),
-            data,
-            &msg_j.psi_prime_prime.0,
-            &security_params.pi_log,
-            &msg_j.psi_prime_prime.1,
-        )
-        .is_err()
-        {
-            faulty_parties.push((j, ciphertext_id, msg_id))
-        }
-    }
-    runtime.yield_now().await;
-
+    let faulty_parties =
+        utils::collect_blame_with_data(&round3_msgs, &ciphertexts, |j, msg_j, ciphertext_j| {
+            pi_elog::non_interactive::verify::<E, D>(
+                &unambiguous::ProofElog {
+                    sid,
+                    prover: j,
+                    prime: true,
+                },
+                pi_elog::Data {
+                    l: &ciphertext_j.A1,
+                    m: &ciphertext_j.A2,
+                    x: &ciphertext_j.Y,
+                    y: &msg_j.Delta,
+                    h: &Gamma,
+                },
+                &msg_j.psi_prime.0,
+                &msg_j.psi_prime.1,
+            )
+            .err()
+        });
     if !faulty_parties.is_empty() {
         return Err(SigningAborted::InvalidPsiPrimePrime(faulty_parties).into());
     }
+    runtime.yield_now().await;
 
     // Step 2
     tracer.stage("Calculate presignature");
     let delta = delta_i + round3_msgs.iter().map(|m| m.delta).sum::<Scalar<E>>();
     let Delta = Delta_i + round3_msgs.iter().map(|m| m.Delta).sum::<Point<E>>();
+    let S = S_i + round3_msgs.iter().map(|m| m.S).sum::<Point<E>>();
```

Comment:
I guess it's the same here as above. The addition is not in the spec, but the sum is incomplete if we don't add `S_i`, right?

Reply:
Yep, exactly, message store only has messages received from other parties